# Document retrieval from wikipedia data

# Fire up GraphLab Create

In [1]:
import graphlab

# Load some text data - from wikipedia, pages on people

In [2]:
people = graphlab.SFrame('people_wiki.gl/')

[INFO] This non-commercial license of GraphLab Create is assigned to fer.gonzalez.rodriguez@gmail.comand will expire on September 21, 2016. For commercial licensing options, visit https://dato.com/buy/.

[INFO] Start server at: ipc:///tmp/graphlab_server-461 - Server binary: /Users/fer_gonzalez_rodriguez/.graphlab/anaconda/lib/python2.7/site-packages/graphlab/unity_server - Server log: /tmp/graphlab_server_1445029271.log
[INFO] GraphLab Server Version: 1.6.1


Data contains:  link to wikipedia article, name of person, text of article.

In [3]:
people.head()

Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

In [4]:
len(people)

59071

# Explore the dataset and checkout the text it contains

## Exploring the entry for president Obama

In [5]:
obama = people[people['name'] == 'Barack Obama']

In [6]:
obama

Columns:
	URI	str
	name	str
	text	str

Rows: Unknown

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

In [7]:
obama['text']

dtype: str
Rows: ?
['barack hussein obama ii brk husen bm born august 4 1961 is the 44th and current president of the united states and the first african american to hold the office born in honolulu hawaii obama is a graduate of columbia university and harvard law school where he served as president of the harvard law review he was a community organizer in chicago before earning his law degree he worked as a civil rights attorney and taught constitutional law at the university of chicago law school from 1992 to 2004 he served three terms representing the 13th district in the illinois senate from 1997 to 2004 running unsuccessfully for the united states house of representatives in 2000in 2004 obama received national attention during his campaign to represent illinois in the united states senate with his victory in the march democratic party primary his keynote address at the democratic national convention in july and his election to the senate in november he began his presidential campa

## Exploring the entry for actor George Clooney

In [8]:
clooney = people[people['name'] == 'George Clooney']
clooney['text']

dtype: str
Rows: ?
['george timothy clooney born may 6 1961 is an american actor writer producer director and activist he has received three golden globe awards for his work as an actor and two academy awards one for acting and the other for producingclooney made his acting debut on television in 1978 and later gained wide recognition in his role as dr doug ross on the longrunning medical drama er from 1994 to 1999 for which he received two emmy award nominations while working on er he began attracting a variety of leading roles in films including the superhero film batman robin 1997 and the crime comedy out of sight 1998 in which he first worked with a director who would become a longtime collaborator steven soderbergh in 1999 clooney took the lead role in three kings a wellreceived war satire set during the gulf warin 2001 clooneys fame widened with the release of his biggest commercial success the heist comedy oceans eleven the first of the film trilogy a remake of the 1960 film wit

# Get the word counts for Obama article

In [9]:
obama['word_count'] = graphlab.text_analytics.count_words(obama['text'])

In [10]:
print obama['word_count']

[{'operations': 1, 'represent': 1, 'office': 2, 'unemployment': 1, 'is': 2, 'doddfrank': 1, 'over': 1, 'unconstitutional': 1, 'domestic': 2, 'named': 1, 'ending': 1, 'ended': 1, 'proposition': 1, 'seats': 1, 'graduate': 1, 'worked': 1, 'before': 1, 'death': 1, '20': 2, 'taxpayer': 1, 'inaugurated': 1, 'obamacare': 1, 'civil': 1, 'mccain': 1, 'to': 14, '4': 1, 'policy': 2, '8': 1, 'has': 4, '2011': 3, '2010': 2, '2013': 1, '2012': 1, 'bin': 1, 'then': 1, 'his': 11, 'march': 1, 'gains': 1, 'cuba': 1, 'californias': 1, '1992': 1, 'new': 1, 'not': 1, 'during': 2, 'years': 1, 'continued': 1, 'presidential': 2, 'husen': 1, 'osama': 1, 'term': 3, 'equality': 1, 'prize': 1, 'lost': 1, 'stimulus': 1, 'january': 3, 'university': 2, 'rights': 1, 'gun': 1, 'republican': 2, 'rodham': 1, 'troop': 1, 'withdrawal': 1, 'involvement': 3, 'response': 3, 'where': 1, 'referred': 1, 'affordable': 1, 'attorney': 1, 'school': 3, 'senate': 3, 'house': 2, 'national': 2, 'creation': 1, 'related': 1, 'hawaii': 1,

## Sort the word counts for the Obama article

### Turning dictonary of word counts into a table

In [11]:
obama_word_count_table = obama[['word_count']].stack('word_count', new_column_name = ['word','count'])

### Sorting the word counts to show most common words at the top

In [12]:
obama_word_count_table.head()

Columns:
	word	str
	count	int

Rows: 10

Data:
+------------+-------+
|    word    | count |
+------------+-------+
|    cuba    |   1   |
| relations  |   1   |
|   sought   |   1   |
|   combat   |   1   |
|   ending   |   1   |
| withdrawal |   1   |
|   state    |   1   |
|  islamic   |   1   |
|     by     |   1   |
|   gains    |   1   |
+------------+-------+
[10 rows x 2 columns]

In [13]:
obama_word_count_table.sort('count', ascending=False)

Columns:
	word	str
	count	int

Rows: 273

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   40  |
|   in  |   30  |
|  and  |   21  |
|   of  |   18  |
|   to  |   14  |
|  his  |   11  |
| obama |   9   |
|  act  |   8   |
|   he  |   7   |
|   a   |   7   |
+-------+-------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Most common words include uninformative words like "the", "in", "and",...

## Summary

### - Corpus = people['text']
### - Training data = table(doc id, doc text)
### - Feature extraction = apply tf-idf(Training data) -> X = tf-idf

<b>obama</b>: 1 record [1 rows x 4 columns]: URI, name, text, word_count (as dict)

<b>obama_word_count_table</b>: table [273 rows x 2 columns]: word, count; descending ORDERED by count

# Compute TF-IDF for the corpus 

To give more weight to informative words, we weigh them by their TF-IDF scores.

In [15]:
people['word_count'] = graphlab.text_analytics.count_words(people['text'])
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
| harpdog brown is a singer ... | {'just': 1, 'issued': 1, '... |
| franz rottensteiner born i... | {'all': 1, 'bauforschung':... |
| henry krvits born 30 decem... | {'they': 1, 'gangstergenka... |
| sam henderson born october... | {'currently': 1, 'less': 1... |
| aaron lacrate is an americ... | {'exclusive': 2, 'producer... |
| trevor ferguson aka john f... | {'taxi': 1, 'salon': 1, 'g... |
| grant nelson born 27 april... | {'houston': 1, 'frankie': ... |
| cathy caruth born 1955 is ... | {'phenomenon': 1, 'deboras... |
+-------------------------------+-------------------------------+
[10 rows x 4 columns]

In [16]:
tfidf = graphlab.text_analytics.tf_idf(people['word_count'])
#tfidf

In [17]:
people['tfidf'] = tfidf['docs']

In [18]:
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
| harpdog brown is a singer ... | {'just': 1, 'issued': 1, '... |
| franz rottensteiner born i... | {'all': 1, 'bauforschung':... |
| henry krvits born 30 decem... | {'they': 1, 'gangstergenka... |
| sam henderson born october... | {'currently': 1, 'less': 1... |
| aaron lacrate is an americ... | {'exclusive': 2, 'producer... |
| trevor ferguson aka john f... | {'taxi': 1, 'salon': 1, 'g... |
| grant nelson born 27 april... | {'houston': 1, 'frankie': ... |
| cathy caruth born 1955 is ... | {'phenomenon': 1, 'deboras... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'selection': 3.8365785530... |
| {'precise': 6.443200606955... |
| {'just': 2.700729968710864... |
| {'all': 1.6431112434912472... |
| {'they': 1.899340117819389... |
| {'currently': 1.6370889691... |
| {'exclusive': 10.455187230... |
| {'taxi': 6.052021456094502... |
| {'houston': 3.935505942157... |
| {'phenomenon': 5.750053426... |
+-------------------------------+
[10 rows x 5 columns]

## Examine the TF-IDF for the Obama article

In [19]:
obama = people[people['name'] == 'Barack Obama']

In [20]:
obama[['tfidf']].stack('tfidf',new_column_name=['word','tfidf']).sort('tfidf',ascending=False)

Columns:
	word	str
	tfidf	float

Rows: 273

Data:
+-------------+---------------+
|     word    |     tfidf     |
+-------------+---------------+
|    obama    | 43.2956530721 |
|     act     |  27.678222623 |
|     iraq    |  17.747378588 |
|   control   | 14.8870608452 |
|     law     | 14.7229357618 |
|   ordered   | 14.5333739509 |
|   military  | 13.1159327785 |
| involvement | 12.7843852412 |
|   response  | 12.7843852412 |
|  democratic | 12.4106886973 |
+-------------+---------------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Words with highest TF-IDF are much more informative.

## Work-flow
  - <b> Training data = table(doc id, doc text)</b>
  - <b> Feature extraction = apply tf-idf(Training data) -> X = tf-idf</b>
  - θ = cluster centers
  - h = estimated cluster label
  - ML model = clustering
  - ML algorithm = k-means
  - Quality metric = distance to cluster centers?


# <b> MANUALLY </b> compute distances between a few people

Let's manually compare the distances between the articles for a few famous people.  

In [21]:
clinton = people[people['name'] == 'Bill Clinton']

In [22]:
beckham = people[people['name'] == 'David Beckham']

In [23]:
#my_test
#clinton

In [24]:
#my_test
#clinton['tfidf'][0]

{'19': 2.8848176417706224,
 '1946': 3.8033836874819134,
 '1977': 2.6962027980008787,
 '1979': 2.6032908378122737,
 '1979born': 10.293348208665249,
 '1981': 2.5978176200443843,
 '1983': 2.57266471080411,
 '1992': 4.556702628633896,
 '1993': 2.3032794139845048,
 '1994': 2.2375557576874794,
 '1998': 4.1375652641876135,
 '2000': 1.8763068991994527,
 '2001': 3.8560499331742757,
 '2004': 1.6903443608359008,
 '2008': 3.0186782749572307,
 '2009': 4.693309450812809,
 '2010': 1.5928339601219734,
 '2012': 1.7938099524877322,
 '2013': 1.9545642372230505,
 '40': 3.490953445340938,
 '42nd': 6.604468754551313,
 '46': 5.230753175638283,
 'a': 0.06743021367099776,
 'acquitted': 5.923900356198227,
 'active': 2.7479584590534256,
 'address': 4.8023464982877115,
 'after': 1.8886668840026128,
 'against': 2.0079609791418744,
 'age': 2.138848033513307,
 'agreement': 4.6496696580787535,
 'aids': 5.131423467022767,
 'also': 0.4627270916162349,
 'alumnus': 4.985080511264044,
 'american': 3.3821333532750204,
 'an

## Is Obama closer to Clinton than to Beckham?

### We will use cosine distance, which is given by  (1-cosine_similarity) 

and find that the article about president Obama is closer to the one about former president Clinton than that of footballer David Beckham.

In [25]:
graphlab.distances.cosine(obama['tfidf'][0],clinton['tfidf'][0])

0.8339854936884277

In [26]:
graphlab.distances.cosine(obama['tfidf'][0],beckham['tfidf'][0])

0.9791305844747478

# AUTOMATICALLY
# Build a nearest neighbor model for document retrieval

We now create a nearest-neighbors model and apply it to document retrieval.  

    + Input: Query article (a document)
    + Corpus: set of docs.
    + Specify: distance metric
    + Output: (list of k) most similar articles

    k - Nearest neighbor (k=1)

        FOR each article 'a' in corpus
            Compute s = similarity(query, a)
            IF (s > best_s) THEN
               add a to output
               best_s = s
            END_IF
        END_FOR

        return output


In [29]:
#my_test
people.head(2)

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 2

Data:
+-------------------------------+----------------+
|              URI              |      name      |
+-------------------------------+----------------+
| <http://dbpedia.org/resour... | Digby Morrell  |
| <http://dbpedia.org/resour... | Alfred J. Lewy |
+-------------------------------+----------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'selection': 3.8365785530... |
| {'precise': 6.443200606955... |
+-------------------------------+
[2 rows x 5 columns]

In [30]:
knn_model = graphlab.nearest_neighbors.create(dataset=people, features=['tfidf'], label='name')

PROGRESS: Starting brute force nearest neighbors model training.


In [31]:
type (knn_model)

graphlab.toolkits.nearest_neighbors._nearest_neighbors.NearestNeighborsModel

# Applying the nearest-neighbors model for retrieval

## Who is closest to Obama?

In [34]:
obama

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 1

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'operations': 1, 'represe... | {'operations': 3.811771079... |
+-------------------------------+-------------------------------+
[1 rows x 5 columns]

In [32]:
knn_model.query(obama)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 35.786ms     |
PROGRESS: | Done         |         | 100         | 532.173ms    |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+----------------+------+
| query_label | reference_label |    distance    | rank |
+-------------+-----------------+----------------+------+
|      0      |   Barack Obama  |      0.0       |  1   |
|      0      |    Joe Biden    | 0.794117647059 |  2   |
|      0      |  Joe Lieberman  | 0.794685990338 |  3   |
|      0      |   Kelly Ayotte  | 0.811989100817 |  4   |
|      0      |   Bill Clinton  | 0.813852813853 |  5   |
+-------------+-----------------+----------------+------+
[5 rows x 4 columns]

As we can see, president Obama's article is closest to the one about his vice-president Biden, and those of other politicians.  

## Other examples of document retrieval

In [37]:
swift = people[people['name'] == 'Taylor Swift']

In [38]:
knn_model.query(swift)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 26.924ms     |
PROGRESS: | Done         |         | 100         | 540.366ms    |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+----------------+------+
| query_label | reference_label  |    distance    | rank |
+-------------+------------------+----------------+------+
|      0      |   Taylor Swift   |      0.0       |  1   |
|      0      | Carrie Underwood | 0.76231884058  |  2   |
|      0      |   Alicia Keys    | 0.764705882353 |  3   |
|      0      |  Jordin Sparks   | 0.769633507853 |  4   |
|      0      |   Leona Lewis    | 0.776119402985 |  5   |
+-------------+------------------+----------------+------+
[5 rows x 4 columns]

In [39]:
jolie = people[people['name'] == 'Angelina Jolie']

In [41]:
knn_model.query(jolie)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 39.897ms     |
PROGRESS: | Done         |         | 100         | 594.266ms    |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------+----------------+------+
| query_label |  reference_label   |    distance    | rank |
+-------------+--------------------+----------------+------+
|      0      |   Angelina Jolie   |      0.0       |  1   |
|      0      |     Brad Pitt      | 0.784023668639 |  2   |
|      0      |   Julianne Moore   | 0.795857988166 |  3   |
|      0      | Billy Bob Thornton | 0.803069053708 |  4   |
|      0      |   George Clooney   |   0.8046875    |  5   |
+-------------+--------------------+----------------+------+
[5 rows x 4 columns]

In [42]:
arnold = people[people['name'] == 'Arnold Schwarzenegger']

In [43]:
knn_model.query(arnold)

PROGRESS: Starting pairwise querying.
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | Query points | # Pairs | % Complete. | Elapsed Time |
PROGRESS: +--------------+---------+-------------+--------------+
PROGRESS: | 0            | 1       | 0.00169288  | 28.796ms     |
PROGRESS: | Done         |         | 100         | 553.879ms    |
PROGRESS: +--------------+---------+-------------+--------------+


Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------------+----------------+------+
| query_label |    reference_label    |    distance    | rank |
+-------------+-----------------------+----------------+------+
|      0      | Arnold Schwarzenegger |      0.0       |  1   |
|      0      |     Jesse Ventura     | 0.818918918919 |  2   |
|      0      |     John Kitzhaber    | 0.824615384615 |  3   |
|      0      |     Lincoln Chafee    | 0.833876221498 |  4   |
|      0      |      Anthony Foxx     | 0.833910034602 |  5   |
+-------------+-----------------------+----------------+------+
[5 rows x 4 columns]